src/artist_22e_pro.cpp

```diff
@@ -0,0 +1,233 @@
+            std::string evstring = std::to_string(eventType);
+            config["mapping"]["buttons"][std::to_string(key)][evstring] = codes;
+        };
+
+        auto addToDialMap = [&config](int dial, int value, int eventType, std::vector<int> codes) {
+            std::string strvalue = std::to_string(value);
+            std::string evstring = std::to_string(eventType);
+            config["mapping"]["dials"][std::to_string(dial)][strvalue][evstring] = codes;
+        };
+
+        // We are going to emulate the default mapping of the device
+        addToButtonMap(BTN_0, EV_KEY, {KEY_B});
+        addToButtonMap(BTN_1, EV_KEY, {KEY_E});
+        addToButtonMap(BTN_2, EV_KEY, {KEY_LEFTALT});
+        addToButtonMap(BTN_3, EV_KEY, {KEY_SPACE});
+        addToButtonMap(BTN_4, EV_KEY, {KEY_LEFTCTRL, KEY_S});
+        addToButtonMap(BTN_5, EV_KEY, {KEY_LEFTCTRL, KEY_Z});
+        addToButtonMap(BTN_6, EV_KEY, {KEY_LEFTCTRL, KEY_LEFTALT, KEY_Z});
+        addToButtonMap(BTN_7, EV_KEY, {KEY_LEFTCTRL, KEY_LEFTSHIFT, KEY_Z});
+        addToButtonMap(BTN_8, EV_KEY, {KEY_V});
+        addToButtonMap(BTN_9, EV_KEY, {KEY_L});
+        addToButtonMap(BTN_SOUTH, EV_KEY, {KEY_LEFTCTRL, KEY_0});
+        addToButtonMap(BTN_EAST, EV_KEY, {KEY_LEFTCTRL, KEY_N});
+        addToButtonMap(BTN_C, EV_KEY, {KEY_LEFTCTRL, KEY_LEFTSHIFT, KEY_N});
+        addToButtonMap(BTN_NORTH, EV_KEY, {KEY_LEFTCTRL, KEY_E});
+        addToButtonMap(BTN_WEST, EV_KEY, {KEY_F});
+        addToButtonMap(BTN_Z, EV_KEY, {KEY_D});
+    }
+    jsonConfig = config;
+
+    submitMapping(jsonConfig);
+}
+
+int artist_22e_pro::sendInitKeyOnInterface() {
+    return 0x02;
+}
+
+bool artist_22e_pro::attachToInterfaceId(int interfaceId) {
+    switch (interfaceId) {
+        case 2:
+            return true;
+        default:
+            return false;
+    }
+}
+
+bool artist_22e_pro::attachDevice(libusb_device_handle *handle, int interfaceId) {
+    unsigned char* buf = new unsigned char[12];
+
+    // We need to get a few more bits of information
+    if (libusb_get_string_descriptor(handle, 0x64, 0x0409, buf, 12) != 12) {
+        std::cout << "Could not get descriptor" << std::endl;
+        return false;
+    }
+
+    int maxWidth = (buf[3] << 8) + buf[2];
+    int maxHeight = (buf[5] << 8) + buf[4];
+    int maxPressure = (buf[9] << 8) + buf[8];
+
+    unsigned short vendorId = 0x28bd;
+    unsigned short productId = 0xf90b;
+    unsigned short versionId = 0x0001;
+
+    struct uinput_pen_args penArgs {
+        .maxWidth = maxWidth,
+        .maxHeight = maxHeight,
+        .maxPressure = maxPressure,
+        .maxTiltX = 60,
+        .maxTiltY = 60,
+        .vendorId = vendorId,
+        .productId = productId,
+        .versionId = versionId,
+        {"XP-Pen Artist 22E Pro"},
+    };
+
+    struct uinput_pad_args padArgs {
+        .padButtonAliases = padButtonAliases,
+        .hasWheel = false,
+        .hasHWheel = false,
+        .wheelMax = 0,
+        .hWheelMax = 0,
```

Comment:
I wasn't sure whether I should remove that entirely or set the values like I did, let me know

Reply:
This looks fine. On older compilers it complains if we don't have them all there in order.